include/data_sink.hpp

```diff
@@ -0,0 +1,900 @@
+        auto            sink = find_sink<T>(query);
+        if (!sink) {
+            return false;
+        }
+
+        sink->template register_multiplexed_callback<O, Callback>(maximum_window_size, std::move(callback));
+        return true;
+    }
+
+    template<typename T, TriggerPredicate P, typename Callback>
+    bool
+    register_snapshot_callback(const data_sink_query &query, P p, std::chrono::nanoseconds delay, Callback callback) {
+        std::lock_guard lg{ _mutex };
+        auto            sink = find_sink<T>(query);
+        if (!sink) {
+            return false;
+        }
+
+        sink->template register_snapshot_callback<P, Callback>(std::forward<P>(p), delay, std::forward<Callback>(callback));
+        return true;
+    }
+
+private:
+    template<typename T>
+    data_sink<T> *
+    find_sink(const data_sink_query &query) {
+        auto matches = [&query](const std::any &v) {
+            try {
+                auto       sink                = std::any_cast<data_sink<T> *>(v);
+                const auto sink_name_matches   = !query.sink_name || *query.sink_name == sink->name();
+                const auto signal_name_matches = !query.signal_name || *query.signal_name == sink->signal_name;
+                return sink_name_matches && signal_name_matches;
+            } catch (...) {
+                return false;
+            }
+        };
+
+        const auto it = std::find_if(_sinks.begin(), _sinks.end(), matches);
+        if (it == _sinks.end()) {
+            return nullptr;
+        }
+
+        return std::any_cast<data_sink<T> *>(*it);
+    }
+};
+
+namespace detail {
+template<typename T, typename P>
+std::span<T>
+find_matching_prefix(std::span<T> s, P predicate) {
+    const auto nm = std::find_if_not(s.begin(), s.end(), predicate);
+    if (nm == s.end()) {
+        return s;
+    }
+    return s.first(std::distance(s.begin(), nm));
+}
+
+template<typename T>
+bool
+copy_span(std::span<const T> src, std::span<T> dst) {
+    assert(src.size() <= dst.size());
+    if (src.size() > dst.size()) {
+        return false;
+    }
+    std::copy(src.begin(), src.end(), dst.begin());
+    return true;
+}
+} // namespace detail
+
+/**
+ * @brief generic data sink for exporting arbitrary-typed streams to non-GR C++ APIs.
+ *
+ * Each sink registers with a (user-defined/exchangeable) global registry that can be
+ * queried by the non-GR caller to find the sink responsible for a given signal name, etc.
+ * and either retrieve a poller handler that allows asynchronous data from a different thread,
+ * or register a callback that is invoked by the sink if the user-conditions are met.
+ *
+ * <pre>
+ * @code
+ *         ╔═══════════════╗
+ *    in0 ━╢   data sink   ║                      ┌──── caller ────┐
+ * (err0) ━╢ (opt. error)  ║                      │                │
+ *         ║               ║  retrieve poller or  │ (custom non-GR │
+ *        ━╢ :signal_name  ║←--------------------→│  user code...) │
```

Comment:
tiny change: replace unused `╢` with `║` 

as discussed: we decided that each sink has only one input at a time. this simplifies the internal design and potential error handling mechanism in case input signals have incompatible sampling rates.

Reply:
Ascii art fix Done